File: lib/domain/politics/government.dart
import '../../core/seed/deterministic_random.dart';

/// Cargos do sistema político do GDD (seção 8).
enum PoliticalOffice {
  governor('Governador', 'Controla impostos, salários públicos e segurança.'),
  districtAdmin('Administrador de Distrito', 'Gere um distrito da capital.'),
  interimGovernor('Governador Provisório', 'Assume após um golpe, até a próxima eleição.'),
  militia('Milícia', 'Braço armado do governo. Reprime rebeliões.');

  const PoliticalOffice(this.label, this.description);
  final String label;
  final String description;
}

/// Governo de uma cidade.
class Government {
  Government({
    required this.settlementId,
    this.governorId,
    this.governorName,
    this.taxRate = 0.08,
    this.publicWage = 40,
    this.treasury = 0,
    this.securityBudget = 0,
    Set<String>? militiaIds,
    Set<String>? wantedIds,
    this.interim = false,
  })  : _militiaIds = {...?militiaIds},
        _wantedIds = {...?wantedIds};

  final String settlementId;

  String? governorId;
  String? governorName;

  /// Imposto sobre transações no Mercado Central. O governador escolhe.
  double taxRate;

  /// Salário diário pago a quem trabalha em Serviços Públicos.
  int publicWage;

  /// Caixa da cidade. É isto que os rebeldes saqueiam num golpe.
  int treasury;

  /// Parcela do tesouro gasta em segurança — soma à força do governo contra
  /// rebeliões e reduz o sucesso de assaltos nas estradas próximas.
  int securityBudget;

  final Set<String> _militiaIds;
  final Set<String> _wantedIds;

  /// `true` quando o cargo veio de um golpe, não de uma eleição.
  bool interim;

  Set<String> get militiaIds => Set.unmodifiable(_militiaIds);
  Set<String> get wantedIds => Set.unmodifiable(_wantedIds);

  bool get hasGovernor => governorId != null;

  static const double maxTaxRate = 0.40;
  static const double minTaxRate = 0.0;

  void setTaxRate(double rate) =>
      taxRate = rate.clamp(minTaxRate, maxTaxRate);

  void collectTax(int amount) => treasury += amount;

  /// Paga salários públicos. Devolve quanto foi efetivamente pago — se o caixa
  /// não cobre a folha, o governo paga o que dá e o resto vira insatisfação.
  int payWages(int workerCount) {
    final due = publicWage * workerCount;
    final paid = due <= treasury ? due : treasury;
    treasury -= paid;
    return paid;
  }

  /// Força defensiva do governo numa rebelião: milícia + dinheiro em segurança.
  double get defenseStrength =>
      _militiaIds.length * 10.0 + securityBudget / 100.0;

  void enlistMilitia(String citizenId) => _militiaIds.add(citizenId);
  void dismissMilitia(String citizenId) => _militiaIds.remove(citizenId);

  void markWanted(String citizenId) => _wantedIds.add(citizenId);
  void pardon(String citizenId) => _wantedIds.remove(citizenId);
  bool isWanted(String citizenId) => _wantedIds.contains(citizenId);

  /// Aplica o resultado de um golpe bem-sucedido: o governador cai, o tesouro é
  /// saqueado e o líder rebelde assume como provisório.
  int applyCoup({required String rebelLeaderId, required String rebelLeaderName}) {
    final looted = treasury;
    treasury = 0;
    governorId = rebelLeaderId;
    governorName = rebelLeaderName;
    interim = true;
    _militiaIds.clear();
    _wantedIds.clear();
    return looted;
  }

  Map<String, dynamic> toJson() => {
        'settlementId': settlementId,
        'governorId': governorId,
        'governorName': governorName,
        'taxRate': taxRate,
        'publicWage': publicWage,
        'treasury': treasury,
        'securityBudget': securityBudget,
        'militiaIds': _militiaIds.toList(),
        'wantedIds': _wantedIds.toList(),
        'interim': interim,
      };

  factory Government.fromJson(Map<String, dynamic> json) => Government(
        settlementId: json['settlementId'] as String,
        governorId: json['governorId'] as String?,
        governorName: json['governorName'] as String?,
        taxRate: (json['taxRate'] as num).toDouble(),
        publicWage: (json['publicWage'] as num).toInt(),
        treasury: (json['treasury'] as num).toInt(),
        securityBudget: (json['securityBudget'] as num).toInt(),
        militiaIds: ((json['militiaIds'] as List?) ?? const [])
            .map((e) => e as String)
            .toSet(),
        wantedIds: ((json['wantedIds'] as List?) ?? const [])
            .map((e) => e as String)
            .toSet(),
        interim: json['interim'] as bool? ?? false,
      );
}

/// Uma candidatura numa eleição.
class Candidacy {
  Candidacy({
    required this.citizenId,
    required this.citizenName,
    required this.platformTaxRate,
    required this.platformWage,
    this.votes = 0,
  });

  final String citizenId;
  final String citizenName;

  /// A plataforma é o que o eleitor avalia: imposto prometido e salário
  /// público prometido.
  final double platformTaxRate;
  final int platformWage;

  int votes;
}

/// Eleição para governador. **Qualquer jogador pode disputar, independente do
/// nível** — é uma regra explícita do GDD.
class Election {
  Election({
    required this.settlementId,
    required this.scheduledForDay,
    List<Candidacy>? candidates,
    this.resolved = false,
    this.winnerId,
  }) : _candidates = [...?candidates];

  final String settlementId;
  final int scheduledForDay;
  final List<Candidacy> _candidates;
  bool resolved;
  String? winnerId;

  List<Candidacy> get candidates => List.unmodifiable(_candidates);

  /// Intervalo entre eleições, em dias.
  static const int termLengthInDays = 30;

  bool register(Candidacy candidacy) {
    if (resolved) return false;
    if (_candidates.any((c) => c.citizenId == candidacy.citizenId)) return false;
    _candidates.add(candidacy);
    return true;
  }

  /// Apura. O eleitorado é simulado: cada eleitor pesa imposto baixo e salário
  /// alto, com um ruído determinístico que representa carisma e Status.
  ///
  /// [statusOf] devolve o atributo Status de um candidato — reputação vira
  /// voto, ligando a economia à política.
  Candidacy? resolve({
    required int electorate,
    required DeterministicRandom rng,
    required int Function(String citizenId) statusOf,
  }) {
    if (resolved || _candidates.isEmpty) return null;

    for (final candidate in _candidates) {
      // Imposto baixo agrada; salário alto agrada; Status amplifica.
      final taxAppeal = (Government.maxTaxRate - candidate.platformTaxRate) /
          Government.maxTaxRate;
      final wageAppeal = (candidate.platformWage / 120).clamp(0.0, 1.5);
      final statusAppeal = statusOf(candidate.citizenId) / 12.0;

      final score = taxAppeal * 0.45 + wageAppeal * 0.35 + statusAppeal * 0.20;
      final noise = rng.rangeDouble(0.85, 1.15);
      candidate.votes = (electorate * score * noise / _candidates.length)
          .round()
          .clamp(0, electorate);
    }

    _candidates.sort((a, b) => b.votes.compareTo(a.votes));
    resolved = true;
    winnerId = _candidates.first.citizenId;
    return _candidates.first;
  }
}

/// Comitê Revolucionário. Se a força rebelde superar a do governo, é golpe de
/// estado; caso contrário, os rebeldes viram procurados no servidor.
class RevolutionaryCommittee {
  RevolutionaryCommittee({
    required this.settlementId,
    required this.leaderId,
    required this.leaderName,
    Map<String, double>? members,
  }) : _members = {...?members};

  final String settlementId;
  final String leaderId;
  final String leaderName;

  /// Id do membro → força de combate que ele traz.
  final Map<String, double> _members;

  Map<String, double> get members => Map.unmodifiable(_members);

  void join(String citizenId, double strength) => _members[citizenId] = strength;
  void leave(String citizenId) => _members.remove(citizenId);

  double get strength => _members.values.fold(0.0, (a, b) => a + b);

  /// Tenta o golpe contra [government].
  CoupResult attemptCoup(Government government) {
    final rebel = strength;
    final loyal = government.defenseStrength;

    if (rebel > loyal) {
      final looted = government.applyCoup(
        rebelLeaderId: leaderId,
        rebelLeaderName: leaderName,
      );
      return CoupResult(
        succeeded: true,
        rebelStrength: rebel,
        governmentStrength: loyal,
        lootedTreasury: looted,
        wantedIds: const [],
      );
    }

    // Fracassou: todo mundo do comitê vira procurado.
    for (final id in _members.keys) {
      government.markWanted(id);
    }
    government.markWanted(leaderId);

    return CoupResult(
      succeeded: false,
      rebelStrength: rebel,
      governmentStrength: loyal,
      lootedTreasury: 0,
      wantedIds: [..._members.keys, leaderId],
    );
  }
}

class CoupResult {
  const CoupResult({
    required this.succeeded,
    required this.rebelStrength,
    required this.governmentStrength,
    required this.lootedTreasury,
    required this.wantedIds,
  });

  final bool succeeded;
  final double rebelStrength;
  final double governmentStrength;
  final int lootedTreasury;
  final List<String> wantedIds;
}

File: web3d/src/politics/government.ts
import type { DeterministicRandom } from '../core/rng';

/**
 * O sistema político da seção 8 do GDD.
 *
 * ## Por que o governo é um objeto de cidade, e não do mundo
 *
 * Cada capital tem o próprio imposto, o próprio salário público e o próprio
 * caixa. É isso que dá razão para migrar: um jogador insatisfeito com o
 * governador não precisa derrubá-lo, pode se mudar. Um governo único global
 * transformaria política em votação anual sem consequência espacial.
 */

export type PoliticalOffice =
  | 'governor'
  | 'districtAdmin'
  | 'interimGovernor'
  | 'militia';

export interface PoliticalOfficeDef {
  readonly id: PoliticalOffice;
  readonly label: string;
  readonly description: string;
}

export const allOffices: readonly PoliticalOfficeDef[] = [
  {
    id: 'governor',
    label: 'Governador',
    description: 'Controla impostos, salários públicos e segurança.',
  },
  {
    id: 'districtAdmin',
    label: 'Administrador de Distrito',
    description: 'Gere um distrito da capital.',
  },
  {
    id: 'interimGovernor',
    label: 'Governador Provisório',
    description: 'Assume após um golpe, até a próxima eleição.',
  },
  {
    id: 'militia',
    label: 'Milícia',
    description: 'Braço armado do governo. Reprime rebeliões.',
  },
];

/**
 * Teto de imposto de mercado — EB 1.1, §14 e §25.
 *
 * Eram 40%, herdados da Rev 3.0. O EB fixa a taxa base em 1% e o teto inicial
 * de governança em 5%, com o argumento de que acima disso o Mercado Central
 * deixa de competir com o clandestino e o governador mata a própria
 * arrecadação sem perceber.
 */
export const MAX_TAX_RATE = 0.05;
export const MIN_TAX_RATE = 0;
/** Intervalo entre eleições, em dias. */
/** Mandato de Governador: 60 dias. A Rev 4.1 revoga os 30 dias anteriores. */
export const TERM_LENGTH_IN_DAYS = 60;

export interface GovernmentJson {
  settlementId: string;
  governorId: string | null;
  governorName: string | null;
  taxRate: number;
  publicWage: number;
  treasury: number;
  securityBudget: number;
  militiaIds: string[];
  wantedIds: string[];
  interim: boolean;
}

export class Government {
  governorId: string | null;
  governorName: string | null;
  /** Imposto sobre transações no Mercado Central. O governador escolhe. */
  taxRate: number;
  /** Salário diário de quem trabalha em Serviços Públicos. */
  publicWage: number;
  /** Caixa da cidade. É isto que os rebeldes saqueiam num golpe. */
  treasury: number;
  /** Parcela gasta em segurança: soma à defesa e reduz assalto nas estradas. */
  securityBudget: number;
  /** `true` quando o cargo veio de um golpe, não de uma eleição. */
  interim: boolean;

  private readonly militia = new Set<string>();
  private readonly wanted = new Set<string>();

  constructor(
    readonly settlementId: string,
    options: Partial<Omit<GovernmentJson, 'settlementId'>> = {},
  ) {
    this.governorId = options.governorId ?? null;
    this.governorName = options.governorName ?? null;
    this.taxRate = options.taxRate ?? 0.01;
    this.publicWage = options.publicWage ?? 20;
    this.treasury = options.treasury ?? 0;
    this.securityBudget = options.securityBudget ?? 0;
    this.interim = options.interim ?? false;
    for (const id of options.militiaIds ?? []) this.militia.add(id);
    for (const id of options.wantedIds ?? []) this.wanted.add(id);
  }

  get militiaIds(): readonly string[] {
    return [...this.militia];
  }

  get wantedIds(): readonly string[] {
    return [...this.wanted];
  }

  get hasGovernor(): boolean {
    return this.governorId !== null;
  }

  setTaxRate(rate: number): void {
    // Teto de 40%: acima disso o Mercado Central deixa de ser competitivo com o
    // clandestino, e o governador consegue matar a própria arrecadação sem
    // perceber. O teto é o que mantém a escolha interessante.
    this.taxRate = Math.min(MAX_TAX_RATE, Math.max(MIN_TAX_RATE, rate));
  }

  collectTax(amount: number): void {
    this.treasury += amount;
  }

  /**
   * Paga a folha pública. Devolve quanto saiu de fato.
   *
   * Caixa curto não trava o pagamento: o governo paga o que dá, e o resto vira
   * insatisfação. Um governo que simplesmente não paga é um governo que o
   * jogador não tem como avaliar.
   */
  payWages(workerCount: number): number {
    const devido = this.publicWage * workerCount;
    const pago = Math.min(devido, this.treasury);
    this.treasury -= pago;
    return pago;
  }

  /** Força defensiva numa rebelião: milícia mais dinheiro em segurança. */
  get defenseStrength(): number {
    return this.militia.size * 10 + this.securityBudget / 100;
  }

  enlistMilitia(citizenId: string): void {
    this.militia.add(citizenId);
  }

  dismissMilitia(citizenId: string): void {
    this.militia.delete(citizenId);
  }

  markWanted(citizenId: string): void {
    this.wanted.add(citizenId);
  }

  pardon(citizenId: string): void {
    this.wanted.delete(citizenId);
  }

  isWanted(citizenId: string): boolean {
    return this.wanted.has(citizenId);
  }

  /**
   * Aplica um golpe bem-sucedido: o governador cai, o tesouro é saqueado e o
   * líder rebelde assume como provisório.
   */
  applyCoup(rebelLeaderId: string, rebelLeaderName: string): number {
    const saqueado = this.treasury;
    this.treasury = 0;
    this.governorId = rebelLeaderId;
    this.governorName = rebelLeaderName;
    this.interim = true;
    this.militia.clear();
    // A lista de procurados também zera: quem o governo anterior perseguia
    // deixa de ser inimigo do governo que acabou de tomar o poder.
    this.wanted.clear();
    return saqueado;
  }

  toJson(): GovernmentJson {
    return {
      settlementId: this.settlementId,
      governorId: this.governorId,
      governorName: this.governorName,
      taxRate: this.taxRate,
      publicWage: this.publicWage,
      treasury: this.treasury,
      securityBudget: this.securityBudget,
      militiaIds: [...this.militia],
      wantedIds: [...this.wanted],
      interim: this.interim,
    };
  }

  static fromJson(json: GovernmentJson): Government {
    return new Government(json.settlementId, json);
  }
}

export interface Candidacy {
  readonly citizenId: string;
  readonly citizenName: string;
  /** A plataforma é o que o eleitor avalia. */
  readonly platformTaxRate: number;
  readonly platformWage: number;
  votes: number;
}

export interface ElectionJson {
  settlementId: string;
  scheduledForDay: number;
  candidates: Candidacy[];
  resolved: boolean;
  winnerId: string | null;
}

/**
 * Eleição para governador.
 *
 * **Qualquer jogador pode disputar, independente do nível** — regra explícita
 * do GDD. É o único caminho de poder que não passa por dinheiro.
 */
export class Election {
  private readonly list: Candidacy[];
  resolved: boolean;
  winnerId: string | null;

  constructor(
    readonly settlementId: string,
    readonly scheduledForDay: number,
    candidates: readonly Candidacy[] = [],
    resolved = false,
    winnerId: string | null = null,
  ) {
    this.list = candidates.map((c) => ({ ...c }));
    this.resolved = resolved;
    this.winnerId = winnerId;
  }

  get candidates(): readonly Candidacy[] {
    return this.list;
  }

  register(candidacy: Omit<Candidacy, 'votes'> & { votes?: number }): boolean {
    if (this.resolved) return false;
    if (this.list.some((c) => c.citizenId === candidacy.citizenId)) return false;
    this.list.push({ ...candidacy, votes: candidacy.votes ?? 0 });
    return true;
  }

  /**
   * Apura.
   *
   * O eleitorado é simulado: cada eleitor pesa imposto baixo e salário alto,
   * com um ruído determinístico que representa carisma. Status entra na conta —
   * reputação vira voto, e é assim que a economia se liga à política.
   */
  resolve(options: {
    electorate: number;
    rng: DeterministicRandom;
    statusOf: (citizenId: string) => number;
  }): Candidacy | null {
    if (this.resolved || this.list.length === 0) return null;

    for (const candidato of this.list) {
      const taxAppeal = (MAX_TAX_RATE - candidato.platformTaxRate) / MAX_TAX_RATE;
      const wageAppeal = Math.min(1.5, Math.max(0, candidato.platformWage / 120));
      const statusAppeal = options.statusOf(candidato.citizenId) / 12;

      const score = taxAppeal * 0.45 + wageAppeal * 0.35 + statusAppeal * 0.2;
      const noise = options.rng.rangeDouble(0.85, 1.15);
      candidato.votes = Math.min(
        options.electorate,
        Math.max(
          0,
          Math.round((options.electorate * score * noise) / this.list.length),
        ),
      );
    }

    this.list.sort((a, b) => b.votes - a.votes);
    this.resolved = true;
    this.winnerId = this.list[0]!.citizenId;
    return this.list[0]!;
  }

  toJson(): ElectionJson {
    return {
      settlementId: this.settlementId,
      scheduledForDay: this.scheduledForDay,
      candidates: this.list.map((c) => ({ ...c })),
      resolved: this.resolved,
      winnerId: this.winnerId,
    };
  }

  static fromJson(json: ElectionJson): Election {
    return new Election(
      json.settlementId,
      json.scheduledForDay,
      json.candidates ?? [],
      json.resolved ?? false,
      json.winnerId ?? null,
    );
  }
}

export interface CoupResult {
  readonly succeeded: boolean;
  readonly rebelStrength: number;
  readonly governmentStrength: number;
  readonly lootedTreasury: number;
  readonly wantedIds: readonly string[];
}

/**
 * Comitê Revolucionário.
 *
 * Se a força rebelde superar a do governo, é golpe. Senão, todo o comitê vira
 * procurado — inclusive quem só entrou para somar número. O risco assimétrico
 * é o que impede a revolução de ser a jogada óbvia todo mês.
 */
export class RevolutionaryCommittee {
  private readonly membros = new Map<string, number>();

  constructor(
    readonly settlementId: string,
    readonly leaderId: string,
    readonly leaderName: string,
    members: Iterable<[string, number]> = [],
  ) {
    for (const [id, forca] of members) this.membros.set(id, forca);
  }

  get members(): ReadonlyMap<string, number> {
    return this.membros;
  }

  join(citizenId: string, strength: number): void {
    this.membros.set(citizenId, strength);
  }

  leave(citizenId: string): void {
    this.membros.delete(citizenId);
  }

  get strength(): number {
    let soma = 0;
    for (const f of this.membros.values()) soma += f;
    return soma;
  }

  attemptCoup(government: Government): CoupResult {
    const rebelStrength = this.strength;
    const governmentStrength = government.defenseStrength;

    if (rebelStrength > governmentStrength) {
      const lootedTreasury = government.applyCoup(this.leaderId, this.leaderName);
      return {
        succeeded: true,
        rebelStrength,
        governmentStrength,
        lootedTreasury,
        wantedIds: [],
      };
    }

    for (const id of this.membros.keys()) government.markWanted(id);
    government.markWanted(this.leaderId);

    return {
      succeeded: false,
      rebelStrength,
      governmentStrength,
      lootedTreasury: 0,
      wantedIds: [...this.membros.keys(), this.leaderId],
    };
  }
}
